docs/design_proposals/port-forward.md

```diff
@@ -0,0 +1,145 @@
+* Status: **Reviewed**/Cancelled/Under implementation/Complete
+
+## Objective
+
+Align Skaffold port-forwarding behaviour with user expectations.
+
+
+## Background
+
+Skaffold supports establishing port-forwarding to deployed Kubernetes resources through
+several mechanisms:
+
+- A user can configure one or more [port forwards in their `skaffold.yaml`](https://skaffold.dev/docs/references/yaml/#portForward)
+  to connect a local port to a remote port on a pod, a service, or some workload resource that
+  has a pod-spec (e.g., Deployment, ReplicaSet, Job).
+- Skaffold automatically configures port forwards for services.
+- When in _debug_ mode (`skaffold debug`), Skaffold automatically configures port-forwards to pods too.
+
+Port forwards are only established when a user runs Skaffold with `--port-forward`.  This was a conscious
+choice in [#969](https://github.com/GoogleContainerTools/skaffold/issues/969) as Skaffold had been
+forwarding all containerPorts defined on pods, and doing so led to conflicts
+in ports chosen.  Users with many services encountered confusion when dealing with many services
+([#1564](https://github.com/GoogleContainerTools/skaffold/issues/1564)).
+
+But putting all port-forwards behind the `--port-forward` flag has resulted in new user confusion
+(e.g., [#4163](https://github.com/GoogleContainerTools/skaffold/issues/4163),
+[#4818](https://github.com/GoogleContainerTools/skaffold/issues/4818)):
+
+> *discr33t* 12:32 PM: I’m having a problem getting Skaffold to port-forward to my service after
+> doing a Helm deploy. If I use the --port-forward flag it will port-forward all my services. But
+> if I define portForward: anywhere in my `skaffold.yaml` nothing is working.
+
+Forwarding all `containerPort`s on pods in `skaffold debug` has also resulted in odd situations
+such as container-ports being allocated before service ports.  For example, the Skaffold
+[`examples/react-reload`](https://github.com/GoogleContainerTools/skaffold/tree/HEAD/examples/react-reload/) example
+has a [service on port 8080](https://github.com/GoogleContainerTools/skaffold/tree/HEAD/examples/react-reload/k8s/deployment.yaml#L7)
+and a [deployment/pod on port 8080](https://github.com/GoogleContainerTools/skaffold/blob/master/examples/react-reload/k8s/deployment.yaml#L29).
+
+
+
+## Design
+
+This document proposes to change Skaffold port-forwarding's defaults in a similar fashion as proposed
+by @corneliusweig in https://github.com/GoogleContainerTools/skaffold/issues/1564#issuecomment-473528574.
+These changes are intended to be backwards-compatible with the previous version
+of Skaffold except where noted.
+
+
+### Command-Line Changes
+
+> **NOTE**: Ideally we would re-purpose Skaffold's existing `--port-forward` CLI argument in
+> a backward-compatible manner (still under investigation).  This document assumes it is
+> not possible.
+
+~Skaffold's `--port-forward` argument should be changed from a binary true/false option to~
```

Comment:
remove since you have a similar note above?

Reply:
Will remove once I've confirmed that this can't be done.